File: Algorithms/C++/695-Max_Area_of_Island.cpp
// 递归 递归深度搜索
// Runtime: 36 ms, faster than 33.98% of C++ online submissions for Max Area of Island.
// Memory Usage: 35.4 MB, less than 6.66% of C++ online submissions for Max Area of Island.
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                res = max(res, area(grid, i, j));
            }
        }
        return res;
    }
    
    int area(const vector<vector<int>>& grid, int row, int col) {
        if (!(0 <= row && row < grid.size() &&
             0 <= col && col < grid[0].size() &&
             seen.find(std::make_pair(row, col)) == seen.end() &&
             grid[row][col] == 1)) {
            return 0;
        }
        seen.insert(std::make_pair(row, col));
        return (1 + 
               area(grid, row + 1, col) +
               area(grid, row - 1, col) +
               area(grid, row, col + 1) +
               area(grid, row, col - 1));
    }
    
private:
    std::set<std::pair<int, int>> seen;
};

File: Algorithms/Python3.x/695-Max_Area_of_Island.py
# Runtime: 172 ms, faster than 13.49% of Python3 online submissions for Max Area of Island.
# Memory Usage: 16.3 MB, less than 60.98% of Python3 online submissions for Max Area of Island.

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.grid = grid
        self.seen = set()
        self.rowLen = len(grid)
        self.colLen = len(grid[0])
        res = 0
        for row in range(self.rowLen):
            for col in range(self.colLen):
                res = max(res, self.area(row, col))
        return res

    def area(self, row, col):
        if 0 <= row < self.rowLen and 0 <= col < self.colLen \
                and (row, col) not in self.seen and self.grid[row][col] == 1:
            self.seen.add((row, col))
            return 1 + self.area(row + 1, col) + self.area(row - 1, col) \
                + self.area(row, col + 1) + self.area(row, col - 1)
        return 0
